File: 257.binary-tree-paths.py
#
# @lc app=leetcode id=257 lang=python3
#
# [257] Binary Tree Paths
#

# Time:  O(n * h)
# Space: O(h)

from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        def helper(node: TreeNode, path):
            if not node.left and not node.right:
                res.append(path + str(node.val))
            if node.left:
                helper(node.left, path + str(node.val) + "->")
            if node.right:
                helper(node.right, path + str(node.val) + "->")

        res = []
        if root:
            helper(root, "")
        return res


# @lc code=end

File: 257.binary-tree-paths.go
/*
 * @lc app=leetcode id=257 lang=golang
 *
 * [257] Binary Tree Paths
 */

package leetcode

import "strconv"

// @lc code=start
func binaryTreePaths(root *TreeNode) []string {
	res := []string{}
	if root != nil {
		helper257(root, "", &res)
	}
	return res
}

func helper257(node *TreeNode, path string, res *[]string) {
	if node.Left == nil && node.Right == nil {
		*res = append(*res, path+strconv.Itoa(node.Val))
	}
	if node.Left != nil {
		helper257(node.Left, path+strconv.Itoa(node.Val)+"->", res)
	}
	if node.Right != nil {
		helper257(node.Right, path+strconv.Itoa(node.Val)+"->", res)
	}
}

// @lc code=end
